pkg/deploy/tls.go

```diff
@@ -0,0 +1,448 @@
+		if !errors.IsNotFound(err) {
+			// Failed to get secret, return error to restart reconcile loop.
+			return false, err
+		}
+
+		// Both secrets (che-tls and self-signed-certificate) are missing which means that we should generate them (i.e. use self-signed certificate).
+		return true, nil
+	}
+	// TLS secret found, consider it as commonly trusted.
+	return false, nil
+}
+
+// GetEndpointTLSCrtChain retrieves TLS certificates chain from given endpoint.
+// If endpoint is not specified, then a test route will be created and used to get router certificates.
+func GetEndpointTLSCrtChain(instance *orgv1.CheCluster, endpointURL string, clusterAPI ClusterAPI) ([]*x509.Certificate, error) {
+	if util.IsTestMode() {
+		return nil, stderrors.New("Not allowed for tests")
+	}
+
+	var requestURL string
+
+	if len(endpointURL) < 1 {
+		var routeStatus RouteProvisioningStatus
+
+		for wait := true; wait; {
+			routeStatus = SyncRouteToCluster(instance, "test", "test", 8080, clusterAPI)
+			if routeStatus.Err != nil {
+				return nil, routeStatus.Err
+			}
+			wait = !routeStatus.Continue || len(routeStatus.Route.Spec.Host) == 0
+			time.Sleep(time.Duration(1) * time.Second)
+		}
+		requestURL = "https://" + routeStatus.Route.Spec.Host
+
+		// Cleanup route after job done.
+		defer func() {
+			logrus.Infof("Deleting a test route %s to extract routes crt", routeStatus.Route.Name)
+			if err := clusterAPI.Client.Delete(context.TODO(), routeStatus.Route); err != nil && !errors.IsNotFound(err) {
+				logrus.Errorf("Failed to delete test route %s: %s", routeStatus.Route.Name, err)
+			}
+		}()
+
+	} else {
+		requestURL = endpointURL
+	}
+
+	// Adding the proxy settings to the Transport object
+	transport := &http.Transport{}
+	if instance.Spec.Server.ProxyURL != "" {
+		logrus.Infof("Configuring proxy with %s to extract crt from the following URL: %s", instance.Spec.Server.ProxyURL, requestURL)
+		configureProxy(instance, transport)
+	}
+	transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	client := &http.Client{
+		Transport: transport,
+	}
+
+	req, err := http.NewRequest("GET", requestURL, nil)
+	resp, err := client.Do(req)
+	if err != nil {
+		logrus.Errorf("An error occurred when reaching test TLS route: %s", err)
+		return nil, err
+	}
+
+	return resp.TLS.PeerCertificates, nil
+}
+
+// GetEndpointTLSCrtBytes creates a test TLS route and gets it to extract certificate chain
+// There's an easier way which is to read tls secret in default (3.11) or openshift-ingress (4.0) namespace
+// which however requires extra privileges for operator service account
+func GetEndpointTLSCrtBytes(instance *orgv1.CheCluster, endpointURL string, clusterAPI ClusterAPI) (certificates []byte, err error) {
+	peerCertificates, err := GetEndpointTLSCrtChain(instance, endpointURL, clusterAPI)
+	if err != nil {
+		if util.IsTestMode() {
+			fakeCrt := make([]byte, 5)
+			return fakeCrt, nil
+		}
+		return nil, err
+	}
+
+	for i := range peerCertificates {
```

Comment:
You can avoid using index of element and simple get element from range: `for _, peerCertificate := range peerCertificates {`